### ISSUE TITLE
Still not right...

### ISSUE THREAD

**Issue by davids5** [thiemar/pyuavcan#2]
@bendyer  -updated everything. This is what I am getting

```
from 

    def decode(self, encoding="utf-8"):
        return bytearray(item.value for item in self.__items
                         if item._bits).decode(encoding)

UnicodeDecodeError: 'utf8' codec can't decode byte 0xff in position 1: invalid start byte
2015-05-09 04:58:23,995 CAN.send(199819021, bytearray(b'\x7f\x08\x00\x00\x00\x00'), True)
2015-05-09 04:58:23,996 CAN.recv(): (200409570, '\x01T\xffn\x06H\x89I', True)
2015-05-09 04:58:49,281 Exception in callback (7, <function null_wrapper at 0x22cf230>)
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/tornado-4.1-py2.7-linux-x86_64.egg/tornado/ioloop.py", line 840, in start
    handler_func(fd_obj, events)
  File "/usr/lib/python2.7/site-packages/tornado-4.1-py2.7-linux-x86_64.egg/tornado/stack_context.py", line 275, in null_wrapper
    return fn(*args, **kwargs)
  File "/home/david_s5/Desktop/dev/PX4/repos/pyuavcan/uavcan/driver.py", line 157, in _read
    callback(self, message)
  File "/home/david_s5/Desktop/dev/PX4/repos/pyuavcan/uavcan/node.py", line 108, in _recv_frame
    h._execute()
  File "/home/david_s5/Desktop/dev/PX4/repos/pyuavcan/uavcan/node.py", line 192, in _execute
    self.on_message(self.message)
  File "/home/david_s5/Desktop/dev/PX4/repos/pyuavcan/uavcan/handlers/__init__.py", line 93, in on_message
    message.unique_id.decode()
  File "/home/david_s5/Desktop/dev/PX4/repos/pyuavcan/uavcan/transport.py", line 311, in decode
    if item._bits).decode(encoding)
  File "/usr/lib64/python2.7/encodings/utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeDecodeError: 'utf8' codec can't decode byte 0xff in position 1: invalid start byte
2015-05-09 04:58:49,284 CAN.send(199819022, bytearray(b'\x99\x08\x00\x00\x00\x00'), True)
2015-05-09 04:58:49,285 CAN.recv(): (200409570, '\x01T\xffn\x06H\x89I', True)
```

**Comment by davids5** [comment]
![image](https://cloud.githubusercontent.com/assets/1945821/7550737/11f47d8c-f609-11e4-9419-2a2cbbc9ef76.png)

**Comment by davids5** [comment]
@bendyer @pavel-kirienko
My uuid
![image](https://cloud.githubusercontent.com/assets/1945821/7551041/047042f2-f615-11e4-861a-e079223a1c66.png)

Note the 0xff in second position

 ame : rx_message_id
    Details:200474377
    Default:200474377
    Decimal:200474377
    Hex:0xbf2ff09
    Binary:1011111100101111111100001001
    Name : rx_len
    Details:6
    Default:6
    Decimal:6
    Hex:0x6
    Binary:110
    Octal:06

 Name : rx_len
    Details:6
    Default:6
    Decimal:6
    Hex:0x6
    Binary:110
    Octal:06
0x20001AEC  00 54 6E 06 48 49 00 00 EF BE AD DE EF BE AD DE EF BE AD DE EF BE AD DE EF BE AD DE EF BE AD DE EF BE AD DE EF BE AD DE EF BE AD DE EF BE AD DE EF BE AD DE EF BE AD DE  .

I tried to add an "ignore" to the decode('utf-8') that defaults to strict - the exception stopped but it will not encode the 0xff

Note the 00 in pos 0
![image](https://cloud.githubusercontent.com/assets/1945821/7551047/76a47cc6-f615-11e4-8567-e0b657683586.png)

What is the correct way to do this?

**Comment by bendyer** [comment]
Just pushed what I hope will be a fix for this — the issue is I've assumed UTF-8 strings everywhere and the hardware UUID needs to be binary. I have therefore added binary string support to transport.ArrayValue, and updated DynamicNodeIDAllocationHandler to use the binary methods rather than the UTF-8 methods.

**Comment by bendyer** [comment]
While investigating that I also ran into an issue with service data type IDs which I've fixed as well. Now both pyuavcan and libuavcan are able to get as far as printing out the GetNodeInfo response.

**Comment by bendyer** [comment]
Fixed in 30244fd
